### PR TITLE
Message interface

### LINKED ISSUE #864: Implement message translator / actor system in shimmer branch
## Description

We have an [implementation](https://github.com/iotaledger/wallet.rs/tree/dev/src/actor) of an actor system (need to rethink about the name) / message passing approach. This needs to be moved to the shimmer branch. 

## Motivation

As we work towards the `shimmer` network, we need to move this implementation to the [shimmer](https://github.com/iotaledger/wallet.rs/tree/shimmer-develop) branch. This is needed so that the bindings can be written for shimmer using this approach. 

## Requirements

- Move actor / message passing approach to [shimmer](https://github.com/iotaledger/wallet.rs/tree/shimmer-develop) branch;
- Document this explaining the motivation, usage so that it is easier for bindings to be written using this approach

## Open questions (optional)

Can the current implementation be improved somehow?

## Are you planning to do it yourself in a pull request?

No.

### PR DESCRIPTION
**Pull request by pelumi-ajayi**
# Description of change

Added message passing interface module for bindings, `actor` module  is now renamed as `message_interface`. Added Objective-C wrapper for Swift binding.

## Links to any relevant issues

Closes #864 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Basic test was done although not exhaustive because interface has not been finalised and can still change. 

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/message_interface/account_method.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2020 IOTA Stiftung
```
Comment (Thoralf-M) [inline src/message_interface/account_method.rs]: Should be 2022, also in the other new files

File: src/message_interface/mod.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+pub use message::Message;
+pub use message_handler::WalletMessageHandler;
+pub use message_type::{AccountToCreate, MessageType};
+pub use response::Response;
+pub use response_type::ResponseType;
+
+use crate::{account_manager::AccountManager, Result};
+use tokio::sync::mpsc::unbounded_channel;
+
+pub async fn create_message_handler(path: Option<String>) -> Result<WalletMessageHandler> {
```
Comment (Thoralf-M) [inline src/message_interface/mod.rs:21]: I think we should also allow to set other manager options here, what do you think about adding Serialize to  `AccountManagerBuilder` and use that with a method `AccountManger::from_str()` or something similar? Would require some change to the SignerHandle, so we can also build it from a String

File: src/message_interface/message_handler.rs [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{AccountToCreate, Message, MessageType, Response, ResponseType, WalletMessageHandler};
+    use crate::account_manager::AccountManager;
+    use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+
+    /// The wallet actor builder.
```
Comment (Thoralf-M) [inline src/message_interface/message_handler.rs]: Some parts still mention `actor`

File: src/message_interface/mod.rs (hunk 2 above)
Comment (pelumi-ajayi) [reply src/message_interface/mod.rs:21]: Right! Setting manager options here makes total sense, I'll play with some of the ideas here but then the SignerHandle changes would have to come from iota.rs.

Comment (Thoralf-M) [reply src/message_interface/mod.rs:21]: Currently for node.js binding we have this, I guess we want something similar
https://github.com/iotaledger/wallet.rs/blob/53d0fe2c2ce8390620f6c1b1f311e46e6ba05a03/bindings/nodejs/src/types.rs#L157
Regarding iota.rs you're right, I can also do that part

Comment (pelumi-ajayi) [reply src/message_interface/mod.rs:21]: Alright thanks, I've made other corrections

Comment (Thoralf-M) [reply src/message_interface/mod.rs:21]: Should be possible with this https://github.com/iotaledger/iota.rs/pull/806

**Pull request merged**